core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java

```diff
@@ -3167,6 +3172,412 @@ public void testRegisterExistingTable() {
+
+    String branchName3 = "retention-branch";
+    int minSnapshotsToKeep = 5;
+    long maxSnapshotAgeMs = Duration.ofDays(7).toMillis(); // 7 days
+    long maxRefAgeMs = Duration.ofDays(30).toMillis(); // 30 days
+
+    table
+        .manageSnapshots()
+        .createBranch(branchName3, firstSnapshot.snapshotId())
+        .setMinSnapshotsToKeep(branchName3, minSnapshotsToKeep)
+        .setMaxSnapshotAgeMs(branchName3, maxSnapshotAgeMs)
+        .setMaxRefAgeMs(branchName3, maxRefAgeMs)
+        .commit();
+
+    // Verify branch with retention policies
+    refs = table.refs();
+    SnapshotRef branchRef = refs.get(branchName3);
+    assertThat(branchRef).isNotNull();
+    assertThat(branchRef.isBranch()).isTrue();
+    assertThat(branchRef.minSnapshotsToKeep()).isEqualTo(minSnapshotsToKeep);
+    assertThat(branchRef.maxSnapshotAgeMs()).isEqualTo(maxSnapshotAgeMs);
+    assertThat(branchRef.maxRefAgeMs()).isEqualTo(maxRefAgeMs);
+
+    table.manageSnapshots().replaceBranch(branchName1, secondSnapshot.snapshotId()).commit();
+
+    // Verify branch points to new snapshot
+    refs = table.refs();
+    SnapshotRef branchRef1 = refs.get(branchName1);
+    assertThat(branchRef1.snapshotId()).isEqualTo(secondSnapshot.snapshotId());
+    assertThat(branchRef1.snapshotId()).isNotEqualTo(firstSnapshot.snapshotId());
+
+    // Add data to feature branch (this should create a new snapshot on the branch)
+    table.newFastAppend().toBranch(branchName1).appendFile(FILE_C).commit();
+
+    // Verify both branches have independent snapshots
+    refs = table.refs();
+    SnapshotRef mainRef = refs.get(SnapshotRef.MAIN_BRANCH);
+    SnapshotRef branchRef1Updated = refs.get(branchName1);
+
+    assertThat(mainRef.snapshotId()).isEqualTo(secondSnapshot.snapshotId());
+    assertThat(branchRef1Updated.snapshotId()).isNotEqualTo(mainRef.snapshotId());
+
+    // Verify the branch snapshot contains FILE_C
+    Snapshot branchSnapshot = table.snapshot(branchRef1Updated.snapshotId());
+    assertThat(branchSnapshot).isNotNull();
+
+    table.manageSnapshots().removeBranch(branchName2).commit();
+
+    // Verify branch is removed
+    refs = table.refs();
+    assertThat(refs).doesNotContainKey(branchName2);
+  }
+
+  @Test
+  public void testTagLifecycleManagement() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).create();
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot firstSnapshot = table.currentSnapshot();
+
+    String tagName1 = "v1.0";
+    table.manageSnapshots().createTag(tagName1, firstSnapshot.snapshotId()).commit();
+
+    // Verify tag was created
+    Map<String, SnapshotRef> refs = table.refs();
+    assertThat(refs).containsKey(tagName1);
+    assertThat(refs.get(tagName1).isTag()).isTrue();
+    assertThat(refs.get(tagName1).snapshotId()).isEqualTo(firstSnapshot.snapshotId());
```

Comment:
Could we make sure these assertions have messages via `as`? Also I feel like a helper like `assertThatRefExists(String ref, RefType type, Long expectedSnapshotId)` would go a long way in reducing the duplication here